SSO integration

lostie opened the pull request:
Trello [ticket](https://trello.com/c/0L7qDexa/323-5-single-sign-on-integration)

This feature includes:
- Adding a link to the login page to enable **SSO** through the existent MOJ SSO application
- Adds configurations to setup the MOJ SSO integration for the different environments
- Assumes **no** multiple SSO providers are supported, only one at a time

Commit:
First draft of SSO integration

Commit:
Add SSO link to the login page

Commit:
Setup SSO integration with the appropriate configurations

For development mode the recommended way of setting up is to have the
SSO running locally and having this application configured with the
appropriate environment variables to correctly connect to the local SSO
application. The environment variables needed to be defined are:
- MOJSSO_ID - the application identifer against the SSO application
- MOJSSO_SECRET - the application's secret to correctly autenticate
against the SSO application
- MOJSSO_URL - the base URL that points to the homepage of the SSO
application

Commit:
Add documentation for SSO integration

Commit:
Ensure login errors are being correctly displayed

The login page contained some flash message in the alert severity which
were not being displayed as there was no setup for it until now.

Ths adds the alert severity to the severities that should be displayed
in the page and add a CSS configuration to display it wit the
appropriate style.

lostie commented:
@alan Could you please review?

alan commented on app/models/user.rb:
```diff
@@ -1,4 +1,16 @@
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.provider = auth.provider
+      user.uid = auth.uid
+      user.email = auth.info.email
```
Have you got a unique index on email? If so you could have the case that someone uses sso after having an account which will result in an error.

lostie replied:
Yes, there's a unique index for the email:

```sql
Indexes:
    "users_pkey" PRIMARY KEY, btree (id)
    "index_users_on_email" UNIQUE, btree (email)
```

What's the recommended strategy to deal with this cases? Check for user by email first and update only the `provider` and `uid`?

alan replied:
Depends on your use case, in PVB `moj-sso` is the only way to sign in so we don't have this scenario. I don't know what is the use case to store the `(provider, uid)` on user...

If you can, making sso the only way to sign in to your app would be the simplest thing to do. For your users (they would only have one account in sso to login) and in the code (you don't even need devise as a dependency).

Commit:
Update user details on SSO user creation for existent user with email

Ensure user creation does not fail when logging in through SSO with a
user that already exists in the local database with the provided e-mail.

This assumes as well that there's no possibility of having multiple SSO
providers (which is the case here, as the only one being used is MOJ
SSO)

lostie commented:
@alan ♻️  with suggested improvements.
